### PR TITLE
Ecc 572

### PR DESCRIPTION
**Pull request by vigneshwaripandi**

### CODE REVIEW

File: app/uk/gov/hmrc/eoricommoncomponent/frontend/controllers/ContactAddressController.scala [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+              BadRequest(contactAddressView(contactAddressDetails, isInReviewMode, formWithErrors, service))
+            },
+          yesNoAnswer => saveAddress().flatMap(_ => locationByAnswer(isInReviewMode, yesNoAnswer, service))
+        )
+    }
+
+  private def fetchContactDetails()(implicit request: Request[AnyContent]): Future[AddressViewModel] =
+    cdsFrontendDataCache.subscriptionDetails flatMap { sd =>
+      sd.contactDetails match {
+        case Some(contactDetails) if contactDetails.street.isDefined =>
```
Comment (yasaracar) [inline app/uk/gov/hmrc/eoricommoncomponent/frontend/controllers/ContactAddressController.scala:78]: Why are we just checking the street here?

File: app/uk/gov/hmrc/eoricommoncomponent/frontend/forms/models/ContactDetailsModel.scala [hunk 2]
```diff
@@ -45,17 +45,13 @@ case class ContactDetailsModel(
...
-    Some(emailAddress),
-    telephone,
-    fax,
-    useAddressFromRegistrationDetails,
-    trim(street),
-    trim(city),
-    trim(postcode),
-    countryCode
-  )
+  def toContactsInfoViewModel: ContactDetailsViewModel =
```
Comment (yasaracar) [inline app/uk/gov/hmrc/eoricommoncomponent/frontend/forms/models/ContactDetailsModel.scala:48]: This should be named as toContactInfoViewModel

File: app/uk/gov/hmrc/eoricommoncomponent/frontend/forms/AddressDetailsForm.scala [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+  def addressDetailsCreateForm()(implicit messages: Messages): Form[AddressViewModel] =
+    Form(
+      mapping(
+        "street"   -> text.verifying(validLine1),
+        "city"     -> text.verifying(validCity),
+        "postcode" -> postcodeMapping,
+        "countryCode" -> text.verifying(
+          "cds.matching-error.country.invalid",
+          s => s.trim.nonEmpty && s != messages("cds.subscription.address-details.country.emptyValueText")
```
Comment (yasaracar) [inline app/uk/gov/hmrc/eoricommoncomponent/frontend/forms/AddressDetailsForm.scala:36]: Why did the validatiopn change here? It was (s => s.length == Length2) before

File: app/uk/gov/hmrc/eoricommoncomponent/frontend/forms/AddressDetailsForm.scala [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        )
+      )(AddressViewModel.apply)(AddressViewModel.unapply)
+    )
+
+  private def validLine1: Constraint[String] =
```
Comment (yasaracar) [inline app/uk/gov/hmrc/eoricommoncomponent/frontend/forms/AddressDetailsForm.scala:41]: Can we rename this to validStreet?

File: app/uk/gov/hmrc/eoricommoncomponent/frontend/services/SubscriptionDetailsService.scala [hunk 5]
```diff
@@ -66,6 +66,30 @@ class SubscriptionDetailsService @Inject() (
       saveSubscriptionDetails(sd => sd.copy(contactDetails = Some(contactDetails)))
     }
 
+  def cacheContactAddressDetails(address: AddressViewModel, contactDetails: ContactDetailsModel)(implicit
+    hc: HeaderCarrier
+  ): Future[Unit] = {
+    val updatedAddress = address.copy(postcode = address.postcode.filter(_.nonEmpty))
```
Comment (yasaracar) [inline app/uk/gov/hmrc/eoricommoncomponent/frontend/services/SubscriptionDetailsService.scala:72]: Why do we have this logic?

File: app/uk/gov/hmrc/eoricommoncomponent/frontend/views/contact_address.scala.html [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+)
+
+
+@(contactAddress: AddressViewModel, isInReviewMode: Boolean, contactAddressDetailsYesNoAnswerForm: Form[YesNo], service: Service)(implicit request: Request[_], messages: Messages)
+
+@fieldsetHint = {
+@govukHint(Hint(id= Some("fieldset-hint"), content = Text(messages("cds.subscription.contact-details.intro")), classes = "govuk-!-margin-bottom-3"))
+}
+
+@labelForEUCountry(countryCode: String) = @{
```
Comment (yasaracar) [inline app/uk/gov/hmrc/eoricommoncomponent/frontend/views/contact_address.scala.html:48]: No need for creating a val here, it is only used once

File: app/uk/gov/hmrc/eoricommoncomponent/frontend/views/contact_address.scala.html [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+).filter(_.nonEmpty)
+}
+
+@hintHtml = {
+<div class="govuk-inset-text" id="contact-address">@{Html(fourLineBusinessAddress.mkString("<br>"))}</div>
+}
+
+
+
+@radioGroupYes = @{
```
Comment (yasaracar) [inline app/uk/gov/hmrc/eoricommoncomponent/frontend/views/contact_address.scala.html:82]: no need for a val here as well

File: app/uk/gov/hmrc/eoricommoncomponent/frontend/views/contact_address.scala.html [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+messages("confirm-contact-address-details.yes")
+}
+
+@radioGroupNo = @{
```
Comment (yasaracar) [inline app/uk/gov/hmrc/eoricommoncomponent/frontend/views/contact_address.scala.html:86]: no need for a val here as well

File: app/uk/gov/hmrc/eoricommoncomponent/frontend/views/contact_address.scala.html [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+messages("confirm-contact-address-details.no")
+}
+
+@displayInputRadioGroup(field: String, options: Seq[(String, String)]) = {
+@inputRadioGroup(
+contactAddressDetailsYesNoAnswerForm(field),
```
Comment (yasaracar) [inline app/uk/gov/hmrc/eoricommoncomponent/frontend/views/contact_address.scala.html:92]: Can you change the alighment of these please?

File: app/uk/gov/hmrc/eoricommoncomponent/frontend/views/contact_address.scala.html [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+    @displayInputRadioGroup("yes-no-answer",
+    Seq("true" -> radioGroupYes,
+    "false" -> radioGroupNo
+    ))
+
+    @govukButton(Button(
+    content = Text(ViewHelper.continueButtonText(isInReviewMode))
+    ))
+    }
+    @helpers.helpAndSupport()
```
Comment (yasaracar) [inline app/uk/gov/hmrc/eoricommoncomponent/frontend/views/contact_address.scala.html:121]: This doesn't have to be in form

**Review by yasaracar — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/eoricommoncomponent/frontend/controllers/ContactAddressController.scala (hunk 1 above)
Comment (vigneshwaripandi) [reply app/uk/gov/hmrc/eoricommoncomponent/frontend/controllers/ContactAddressController.scala:78]: As the Contact Address information is amended to cache only after the address entry page, there is a check needed if any mandatory Address data exists already (here street is checked)  and accordingly retain values from ContactDetails cache to copy address data over to the Review and Submit method of contact-details page

File: app/uk/gov/hmrc/eoricommoncomponent/frontend/forms/models/ContactDetailsModel.scala (hunk 2 above)
Comment (vigneshwaripandi) [reply app/uk/gov/hmrc/eoricommoncomponent/frontend/forms/models/ContactDetailsModel.scala]: This has been changed to toContactInfoViewModel, and related references to this has been updated.

File: app/uk/gov/hmrc/eoricommoncomponent/frontend/views/contact_address.scala.html (hunk 9 above)
Comment (vigneshwaripandi) [reply app/uk/gov/hmrc/eoricommoncomponent/frontend/views/contact_address.scala.html]: Alignment had been fixed now

File: app/uk/gov/hmrc/eoricommoncomponent/frontend/views/contact_address.scala.html (hunk 7 above)
Comment (vigneshwaripandi) [reply app/uk/gov/hmrc/eoricommoncomponent/frontend/views/contact_address.scala.html:82]: The val has been removed

File: app/uk/gov/hmrc/eoricommoncomponent/frontend/views/contact_address.scala.html (hunk 6 above)
Comment (vigneshwaripandi) [reply app/uk/gov/hmrc/eoricommoncomponent/frontend/views/contact_address.scala.html:48]: The val has been removed

File: app/uk/gov/hmrc/eoricommoncomponent/frontend/forms/AddressDetailsForm.scala (hunk 4 above)
Comment (vigneshwaripandi) [reply app/uk/gov/hmrc/eoricommoncomponent/frontend/forms/AddressDetailsForm.scala:41]: This has been now renamed to validStreet

File: app/uk/gov/hmrc/eoricommoncomponent/frontend/forms/AddressDetailsForm.scala (hunk 3 above)
Comment (vigneshwaripandi) [reply app/uk/gov/hmrc/eoricommoncomponent/frontend/forms/AddressDetailsForm.scala]: The validation was used from previous Address Details page entry form, now this has been changed back to validation before as the Length check (s => s.length == Length2)

File: app/uk/gov/hmrc/eoricommoncomponent/frontend/views/contact_address.scala.html (hunk 10 above)
Comment (vigneshwaripandi) [reply app/uk/gov/hmrc/eoricommoncomponent/frontend/views/contact_address.scala.html:110]: The form has been aligned.

File: app/uk/gov/hmrc/eoricommoncomponent/frontend/services/SubscriptionDetailsService.scala (hunk 5 above)
Comment (vigneshwaripandi) [reply app/uk/gov/hmrc/eoricommoncomponent/frontend/services/SubscriptionDetailsService.scala:72]: this logic is needed to copy the address entry page values to the single contact details cache, we are merging the address details into the contact detail cache data.

**Review by yasaracar — APPROVED** [review]

**Pull request merged**